### ISSUE TITLE
pyright is confused by Literal['*'] when the Literal symbol was imported from a module that is not "typing"

### ISSUE THREAD

**Issue by zzzeek** [microsoft/pyright#3025]
pyright 1.1.220

given a package:

```
package/
              __init__.py
              file1.py
              file2.py
```

pyright seems to get confused about `Literal['*']` when `Literal` is imported from a different file:

**file1.py**

```py
from typing import Literal
```

**file2.py**

```py
from .file1 import Literal
from typing import Union

# uncomment to resolve error
# from typing import Literal

def do_thing(x: Union[Literal["*"], Literal[1]]):
    pass
```

pyright output:

```
/home/classic/tmp/foo/package/file2.py
  /home/classic/tmp/foo/package/file2.py:7:31 - error: Unpack operator not allowed in type annotation
  /home/classic/tmp/foo/package/file2.py:7:31 - error: Expected expression
2 errors, 0 warnings, 0 informations 
Completed in 0.688sec
```

We support Python 3.7, 3.8 so we have a compatibility module "sqlalchemy.util.typing" that uses either ``typing.Literal`` or ``typing_extensions.Literal``, which is why we don't have the ``typing.Literal`` import directly in the file that it's used.

**Comment by JelleZijlstra** [comment]
Oh interesting, that probably means it's treating the `"*"` as a forward reference and tries to parse it as PEP 646 syntax.

**Comment by zzzeek** [comment]
that's what the error message says yup

**Comment by erictraut** [comment]
Yes, your supposition is correct. Pyright's parser and binder have some knowledge of a handful of special forms. For completeness, this list includes the following:

* Parser: Literal, TypeAlias, Annotated
* Binder: Final, ClassVar, Required, NotRequired

Aliasing these special forms is therefore not supported. You need to import them directly from either `typing` or `typing_extensions` if you want them to work with pyright.

I could theoretically work around this limitation, but it would involve a lot of complexity and would have a pretty significant impact on analysis performance, so I don't think such a change would be in the the best interest of users.

@zzzeek, it sounds like you've found a good workaround.

**Comment by zzzeek** [comment]
> Aliasing these special forms is therefore not supported. You need to import them directly from either `typing` or `typing_extensions` if you want them to work with pyright.

OK so, how about pyright emits a warning when it sees aliases of these special forms, e.g. "note: alias of Literal not supported, typing may not be accurate" ?  I would think this would help you also so that these known limitations don't look like strange bugs, and instead are noted "hey, we dont really support this" , so that the subsequent misinterpretation of the type is not unexpected.

**Comment by erictraut** [comment]
I'll consider that if it comes up frequently enough.
